Pelt simulator

derkling opened the pull request:
Here is an initial implementation for a simple PELT simulator which allows to study the behaviours of the PELT signal, depending on its configuration parameters, for the tracking of the utilization of periodic tasks.

A simple Notebook to test this new API is available here:
https://gist.github.com/derkling/44c5b75788cb84cb5bedc697e2687305

derkling committed:
sched: add first simple PELT simulator

This patch provide a first implementation of a simple PELT simulator.
Despite being of a bare simplicity it's still a useful tool to show and
study how PELT describes a task depending on its profile (period and
duty-cycle) as well as some specific PELT attributes (half-life, initial
values).
The current PELT model can be easily extended to evaluate experimental
features such as (for example) the capping of the decay value.

Signed-off-by: Patrick Bellasi <patrick.bellasi@arm.com>

derkling commented:
The plotting functions provided by the example Notebook are probably going to be pushed for integration in TRAPpy (http://github.com/ARM-software/trappy) as part of a new analysis API.

derkling commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+# Task Model
+################################################################################
+
+class Task(object):
```
Dunno if it can be worth to have the tasks model definitions in a separate model.
At the end these tasks depends on how the PELT simulator works, that's why I've kept them on the same module.

sinkap replied:
In this case wouldnt it be better to rename it to `SimulatedTask`?

derkling replied:
Not completely convinced, it seems kind-of implied since it's a simulator... but let's make it more clear by renaming it to SimulatedTask

derkling commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        self.duty_cycle_pct = int(100. * run_ps / period_ps)
+        self.util_avg =  int(float(Task.pelt_max) * self.duty_cycle_pct / 100)
+
+    def running(self, time_ms):
```
We need to extend this method to provide a more generic implementation where the input waveform is not necessary periodic. Moreover, we should return the "percentage" of time spent since the last PELT sample as a base for supporting more generic (i.e. non periodic tasks)

sinkap replied:
can you change the name of this method to `isRunning`

derkling replied:
don't like completely... but it sound more English correct ;-)

derkling commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
```
Ok... the time is always wrong... but perhaps this time should be the type :-)

sinkap replied:
:)

derkling commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        self._geom_y = pow(0.5, 1./half_life_ms)
+        self._geom_u = float(Simulator._signal_max) * (1. - self._geom_y)
+
+        print "PELT Configured with   :"
```
Not sure what's the strategy you suggest for the initialization code. I like to get this report but... perhaps we can remove it from here and get it on demand, if the client code want it, using the **str** method defined below...

derkling reviewed (commented):
I've added some comments related to things to perhaps improve before merge.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        run_ps = period_ps * float(duty_cycle_pct) / 100
+        super(PeriodicTask, self).__init__(period_ps, run_ps, start_ps)
+
+class AndroidTask(Task):
```
Why do we call this Android task? and also why is this needed? Can the user not just create a `PeriodicTask` with the parameters they need?

derkling replied:
It was a simplification for our Android friends, but... I agree that it  has not the standing of an API citizen...
Technically the SimulatedPeriodicTask is also just a SimulatedTask where we specify the duty-cycle instead of the run_time.... I'll try also to remove the PeriodTask.

derkling replied:
There is now just on Task model which is named (just) "PeriodicTask" to better call out that we currently support only this simple class of tasks

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        self._geom_y = pow(0.5, 1./half_life_ms)
+        self._geom_u = float(Simulator._signal_max) * (1. - self._geom_y)
+
+        print "PELT Configured with   :"
```
Yeah. Better to add a **str** method. Adding print in the code would not be nice.

derkling replied:
Done

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+
+        # Intervals (in PELT samples) for the signal to compute
+        start_us = int(1e6 * Simulator._sample_us/1e3 * start_s)
+        end_us   = int(1e6 * Simulator._sample_us/1e3 * end_s)
```
Can we refactor the constant divisions to use internal utility functions (module scoped)

`_us_to_ms` and so on. The constant divisions and multiplications seem to be a bit confusing.

derkling replied:
Ok, I've added a new section at the bottom with utility functions to support these units conversions.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        while t_us <= end_us:
+
+            # Check if the task was running in the current PELT sample
+            running = task.running(float(t_us/1e3))
```
constant divisions. Refactor to functions.

derkling replied:
This is also using utility conversion functions

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+"""
+
+from collections import namedtuple as _namedtuple
+from pandas import DataFrame as _DataFrame
```
can you do 

```
import pandas as pd
```

and use pd.DataFrame. Any reasons for using `_DataFrame` here?

JaviMerino replied:
that also struck me as unnecessarily odd.  What's wrong with `pd.DataFrame`?  Furthermore, what's wrong with `namedtuple`?

derkling replied:
Playing with the auto-completion inside iPython Notebooks, I've noticed that everything imported from a module which is not "private" to that module (i.e. starting with an "_"), is available for TAB based auto-completion.
Dunno if this is a "defect" of Notebooks autocompletion... or just an (in)sane behaviour of Python which assumes public everything imported from a module.

All that considered, from a user standpoint, I would like (and find it quite useful) that if she writes:

``` python
peltsim = pelt.Simulator(...)
peltsim.<TAB>
```

than that <TAB> does not expand into a list which includes "namedtuple" and "DataFrame" among the possible options.

I understand that what I did is kind of introducing a "new convention" in the BART code-base, but I would like to know what you guys think about it. Is it ok for you (if the issue is there) to try simplify usage for end-users by hiding within each module everything do not make sense to expose by TAB-completion?

JaviMerino replied:
This is normal for python, it has nothing to do with ipython.  Take any module that you import and do a dir().  You will probably see most of the modules it imports.

In the example that you use, 

``` python
peltsim = pelt.Simulator(...)
peltsim.<TAB>
```

In this case, it will autocomplete for methods defined in the Simulator class, not methods defined in the `pelt` module, so you won't have any problems.

The example you were looking for is:

``` python
from bart.sched import pelt
pelt.<TAB>
```

That would autocomplete imported stuff, like for most python modules.  If that really bothers you there are ways of controlling what is available there, but the `_` is not the right solution.

derkling replied:
Right, I've got confused and I was actually TAB completing pelt (not peltsim)... I'll switch back to the standard names for these external modules. ;-)

derkling replied:
Fixed: using now namedtuple and DataFrame.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+                        self.run_us/1e3, self.duty_cycle_pct,
+                        self.util_avg)
+
+class PeriodicTask(Task):
```
and `SimulatedPeriodicTask`

derkling replied:
ok... names are becoming quite long that way... however, let's rename them.
Otherwise we should assume that a Task is a "SimulatedTask" and get ride of this "abuse of simulation"... which, you know, it is not always good ;-)

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+            u_1 += self._geom_u
+        return u_1
+
+    def stable_range(self, task):
```
we are using the convention of camelcased function names in BART (see other modules). Can use the same? 

`stableRange`

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+                         Simulator._signal_max * max_util)
+
+
+    def signal(self, task, start_s=0, end_s=10):
```
signal can be a verb as well. I think in this case we do need a get

so `getSignal`

derkling replied:
Ok

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+# PELT Model
+################################################################################
+
+PELT_Sim = _namedtuple('PELT_Sim', [
```
We can just name these variables in a camel cased way.

`PeltStats`

or `PELTStats`

whichever you prefer.

derkling replied:
Done... I prefer the second form.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        :type  stats_start_s: float
+
+        :param stats_end_s: the end of signal portion to build the stats for
+        :type  stats_end_s: float
```
Please add a returns to this docstring as well.

derkling replied:
Done. Since we return a PELTStats namedtuple I've also made this symbol visible outside the module and added  a proper docstring to it ;-)

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+
+        return self._df
+
+    def stats(self, task, stats_start_s=0, stats_end_s=None):
```
`getStats` because this does not make it clear if we are printing or just getting the stats.

derkling replied:
ok, updated

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+# limitations under the License.
+#
+
+"""
```
We should have a better docstring for the module. Also, have a look at https://www.python.org/dev/peps/pep-0008/#documentation-strings for the convention on docstrings.

derkling replied:
Added a more detailed description.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+    'tmin', 'tmax', 'min', 'max', 'avg', 'std',
+    'err', 'err_pct'])
+
+PELT_Range = _namedtuple('PELT_Range', [
```
Also, these are module local variables. You can prefix them with an underscore to indicate that they should not be mutated with global imports. Python being python you cannot really enforce this but yes would help with readability.

Have a look at:

https://www.python.org/dev/peps/pep-0008/#global-variable-names

So `_PELTStats` would be a more befitting name.

derkling replied:
Done, while we keep the non "_" name for the namedtuple, i.e.:
```pyhon
_PELTStats = _namedtuple('PELTStats', ...)
```

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+            t_us += Simulator._sample_us
+
+        # Create DataFrame from computed samples
+        self._df = _DataFrame(samples, columns=['Time', 'PELT_Interval', 'Running', 'util_avg'])
```
Also, I am not sure why are we using `util_avg` should it not be just `signal_value`. PELT can theoretically represent any thing. Some of the things being `load_avg` and `util_avg`

derkling replied:
mmmm... right, I've updated to be a more generic "pelt_value" for the DataFrame, while it's "pelt_avg" for the task, since this actually represents the average value we expect from the PELT oscillations.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        #print "Statistics Time Range : ({:.6f}:{:.6f}) [s]".format(stats_start_s, stats_end_s)
+
+        # Compute stats
+        _util_avg = task.util_avg
```
`_util_avg` -> something more generic

derkling replied:
Ad discussed before: this is now pelt_avg

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        while t_us <= end_us:
+
+            # Check if the task was running in the current PELT sample
+            running = task.running(float(t_us/1e3))
```
refactor as function (as above)

derkling replied:
ditto as above...

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        # Add initial value to the generated signal
+        util = self.init_util
+        t_us = start_us
+        samples.append((float(t_us)/1e6, 0, False, util))
```
ditto.

derkling replied:
yelp... done

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+            raise ValueError("Wrong time for task parameter")
+
+        # Compute max value
+        max_util  = (1. - pow(self._geom_y, float(task.run_us)/Simulator._signal_max))
```
Would it be better to have separate functions for max and min values? We can do this in a later CL if you like. but it could be useful I think..

derkling replied:
Well... the stableRange function is already returning a min,max pair... perhaps separate functions for these two values are a bit overkilling?
Moreover, we should consider that, outside of the stability range, one of the two values is always saturated... IOW, min/max make sense only within a stability range.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+    pelt_sample_us = 1024
+    pelt_max = 1024
+
+    def __init__(self, period_ps, run_ps, start_ps=0,
```
Sorry what is `_ps` here? picoseconds?

bjackman replied:
Periods? If so maybe `_pds` would be a better suffix (and mention in docstring)

derkling replied:
"ps" is the short for "PELT Samples"... the simulator uses 1024us as a unit of length to specify tasks properties while keeping computations simple.

bjackman replied:
IMO it's too easy to get that confused with picoseconds.

msrasmussen replied:
I thought it was picoseconds until I read the code carefully. I agree with bjackman it should be either spelled out or the suffix should make it very clear that it isn't seconds/microseconds/picoseconds.

sinkap replied:
I agree with that. Why not just call it `num_period_samples`? and `num_run_samples`

derkling replied:
Agree, it's confusing... and I kind-of like also to @sinkap proposal but to keep the name slightly shorter while still not compromising readability, I'm using `run_samples` and `period_samples`... the documentation clarify what a "sample" is.

sinkap commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,347 @@
+        interval used by the Linux scheduler to update a
+        PELT signal.
+
+        :param start_ps: task start time
```
task start time in....

derkling replied:
Added unit of measure [pelt_sample_us] to all parameters

sinkap reviewed (changes requested):
I have not looked at the decay bits yet. This looks good, a few minor comments.

We will need tests for this as well.

JaviMerino commented:
I don't think it is a good idea to push the plotting into trappy.  I started by adding specific plotters to trappy and in hindsight it was a bad idea.  The trace analysis classes in lisa's env work a lot better, the specific plotting that we currently have in trappy looks completely out of place and I would rather not add more.

derkling commented:
An alternative option is of course to have everything which is a "static" plotting support being available only within LISA... but let's talk this better in the sync hangout this afternoon ;-)

sinkap commented:
I agree. I want to extend the interactive plotting interface more so that
we can have interactive plots in LISA which are area specific and keep
Trappy as a core non specific library.

On Mon, Nov 14, 2016 at 11:23 AM, Patrick Bellasi notifications@github.com
wrote:

> An alternative option is of course to have everything which is a "static"
> plotting support being available only within LISA... but let's talk this
> better in the sync hangout this afternoon ;-)
> 
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> https://github.com/ARM-software/bart/pull/65#issuecomment-260299123, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/ACDZbLrntV5pPOqznYph_u1trMSiz6hxks5q-DaggaJpZM4Ku_dK
> .

sinkap commented:
Let's finish this off. @derkling do you know when you will be able to find time for this?

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        self.period_us = Task.pelt_sample_us * period_ps
+        self.run_us    = Task.pelt_sample_us * run_ps
+
+        self.pelt_sample_us = pelt_sample_us
```
Doesn't seem right that this gets set _after_ the values above - that would make the values nonsense if you override `pelt_sample_us`.

derkling replied:
That's right, I would probably better expose these constants as default values for ctor parameters...
Should that be better?

bjackman replied:
They are already defaults for ctor parameters. But seems to me `self.pelt_sample_us` should be set before `self.start_us`, and then we should reference `self.pelt_sample_us` instead of `Task.pelt_sample_us`

derkling replied:
Right... this seems like it was a WIP mod...  fixed now ;-)

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        # The 1us "before time" is always represeting the exact status of the
+        # task in that period. This is why we remove 1 in the following
+        # computation:
+        period_delta_us = (float(time_ms) * 1e3 - 1) % self.period_us
```
IIUC this solves the problem that calling `.running` with the exact moment 
at which the task "wakes up" would give the appearance of an extra PELT period of runtime. But doesn't this just move the problem to the other end of a task activation i.e. we now get an extra period at the end? Example:

```
In [5]: t = Task(period_ps=10, run_ps=2)

In [12]: t.running(10.240) # Task not shown as running at wakeup
Out[12]: False

In [16]: t.running(10.240 + 2.048)
Out[16]: True

In [17]: t.running(10.240 + 2.047)
Out[17]: True

In [18]: t.running(10.240 + 2.048)
Out[18]: True

In [19]: t.running(10.240 + 2.049) # ^ Extra period at end of activation
Out[19]: True

In [20]: t.running(10.240 + 2.050)
Out[20]: False
```

derkling replied:
[12] is as expected, when we query for running we are looking at the **previous full 1024us window**, thus at wakeup time the task was **not** running in the previous window.

[19] is also true **but not dangerous** since the PELT simulator class, for the time being, will never check +2.049.
Indeed, the simulator generates new values only at PELT samples time instants, thus it will check at +1024, +2048 and +3062.

That's also the reason why task's period and run can be specified only in units of 1024us, it's a limitation of the current model but it also makes it quite simple and still useful for some analysis.

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        Here we use:
+        - Decay factor:       y = 0.5^(1/half_life)
+        - Max stable value:       (1-y^r) / (1-y^p)
+        - Min stable value: y^i * (1-y^r) / (1-y^p)
```
Would be useful to specify what r, p and i are (r = runtime, p = period, i = p - r ?)

derkling replied:
Right

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
```
s/time/type/? Anyway, I don't think we should check types of parameters - if it looks like a Task, talks like a Task and has all the attributes of a Task let's just call it a Task IMO

derkling replied:
I agree on the "looks like a Task, talks like a Task and has all the attributes of a Task let's just call it a Task"... that's why I check for the base class. That's the only why to ensure that we ""looks like a Task"... isn't it?

Otherwise we can have funny users which play by passing in some other object and than complaining for exceptions of kind "your task as not a run_samples attribute"...

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        self.half_life_ms = half_life_ms
+        self.decay_cap_ms = decay_cap_ms
+
+        self._geom_y = pow(0.5, 1./half_life_ms)
```
Can we just use `**` instead of `pow`?

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+
+        # Intervals (in PELT samples) for the signal to compute
+        start_us = int(1e6 * Simulator._sample_us/1e3 * start_s)
+        end_us   = int(1e6 * Simulator._sample_us/1e3 * end_s)
```
These calculations are wrong. AFAICT it should be
```
# Intervals (in PELT samples) for the signal to compute
start_us = (start_s * 1e6) / sample_us
```
(in which case the variables should be renamed) or 
```
# Time (in us) for the signal to compute
start_us = start_s * 1e6
```
Currently the range of `.signal` is wrong:
```
In [9]: df = s.signal(t, end_s=1)

In [10]: df.index[-1]
Out[10]: 1.024
```
(only slightly wrong, because it happens that `Simulator._sample_us/1e3` ~= 1)

derkling replied:
Is not that the intended behaviour? 
One of the simplifications we do is that we compute PELT values only at 1024 [us] intervals... thus, we implicitly translate every [s] time used by the user (e.g. end_s=1) into the nearest non smaller multiple of PELT simulator samples lenght (i.e. 1024 us)

derkling replied:
I'm adding a :note: section to the documentation

bjackman replied:
When `end_s=1`, the nearest non smaller multiple of the sample length is 1.000448s (= 977 * 0.001024s), not 1.024.

derkling replied:
Ok, right... that's a good one! Thanks for spotting it! Going to send a refresh of that code!

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        self.duty_cycle_pct = int(100. * run_ps / period_ps)
+        self.util_avg =  int(float(Task.pelt_max) * self.duty_cycle_pct / 100)
+
+    def running(self, time_ms):
```
Time parameter in uS? No mS anywhere else here.

derkling replied:
No, I'm using a more user friendly [ms] unit for the external  APIs... while [us] are used internally for all the computations. I think that should be acceptable at least from a user standpoint... don't you?

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        :type  start_s: float
+
+        :param end_s: the end time in [s]
+        :type  end_s: float
```
Can you add a `:returns:`? Would be useful to specify the names of the DataFrame columns

derkling replied:
Right, adding a description of the returned dataframe

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+    def stats(self, task, stats_start_s=0, stats_end_s=None):
+        """
+        Compute the stats over a pre-computed PELT signal, considering only
+        the specified portion of the signal.
```
Maybe just me but I don't like the taste of the "precomputed signal" thing, it seems like an unnecessary inflexibility/global state. Why not just take the DataFrame as a parameter rather than using `self._df`?

derkling replied:
Well... than we should probably also avoid to use a class and expose only methods...
I agree which is a quite limited and simple class right now, which makes encapsulation a bit overkilling... but as an overall design for future changes I would prefer to keep an internal state which represents the last computed signal.

What is inconsistent is eventually the fact that we pass Task to both getSignal and getStats... and they can potentially we different objects. Thus, I would propose to remove Task from getStats and keep track of the task we computed the signal for in the getSignal.

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+    pelt_sample_us = 1024
+    pelt_max = 1024
+
+    def __init__(self, period_ps, run_ps, start_ps=0,
```
BTW up to you if you think it's worth validating but there's a divison by zero if period is 0. Also if `run_ps` > `period_ps` we get nonsense util_avg. Both of these are invalid parameters but might be worth `raise`ing to make those errors more obvious to users?

derkling replied:
Right, I'm adding a minimal sanity check for reasonable values...

msrasmussen commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+        min_util  = (max_util * pow(self._geom_y, float(task.idle_us)/Simulator._signal_max))
+
+        return PELT_Range(Simulator._signal_max * min_util,
+                         Simulator._signal_max * max_util)
```
Simulator._signal_max should be replaced with Simulator._sample_us in all of the computations above. We are converting us into PELT samples, hence we should use _sample_us. _signal_max is something completely different, but happens to have the same default value.

derkling replied:
Right, fixed.

sinkap commented:
Thanks for your review @bjackman 👍

msrasmussen commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+
+            # Keep track of sleep start and decay capping time
+            if self.decay_cap_ms and running_prev and not running:
+                capping = t_us + (self.decay_cap_ms * 1e3)
```
Everything else is assume to be aligned with the PELT 1024us sample boundaries and has to be specified in PELT samples, so why isn't this the case for decay_cap_ms? I would argue it has to be aligned as well, otherwise we will get errors. You would want to multiply with _sample_us here instead.

msrasmussen commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,377 @@
+                util = self._geom_sum(util, active_us)
+
+            # Update decay only up to the capping point
+            elif capping and t_us < capping:
```
Shouldn't this be t_us <= capping? Otherwise you clamp one sample too early?

derkling replied:
Mmmm... right!

derkling committed:
sched: add support for decay capping to the PELT simulator

Once a task is sleeping we can cap the decay of its PELT signal to avoid
an underestimation of the task at wakeup time. This patch adds the
required support to the simple PELT simulator to investigate the
behaviours of the PELT signals when different capping times are
configured.

Signed-off-by: Patrick Bellasi <patrick.bellasi@arm.com>

derkling reviewed (commented):
Thanks for the detailed review comments... I think I've addressed all of them at my best.
Let's have one more review cycle and then, next week, I'll add (or maybe @bjackman will add) a test case patch to this series before merging. ;-)

A refreshed version of the test notebook has been pushed at the same link:
https://gist.github.com/derkling/44c5b75788cb84cb5bedc697e2687305

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+
+        args = [period_samples, start_sample, pelt_sample_us, pelt_max]
+        invalid = any([True for param in args if
+                       not isinstance(param, int) or param < 0])
```
```
        invalid = any(not isinstance(param, int) or param < 0 for param in args)
```

derkling commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,464 @@
+# Utility Functions
+################################################################################
+
+def _s_to_us(time_s, interval_us=1e3):
```
@bjackman this should now be a better implementation of the time conversion functions which addresses this previous comments of yours:
https://github.com/ARM-software/bart/pull/65#discussion_r90434299

Have a look and let me know if the usage of the `interval_us` parameter is good enough instead of using a different variable name. The function is still named _s_to_us, but the default "units" can be tuned at call time to be 1024 instead of the default 1e3.

bjackman replied:
Looks good, but the "not smaller" only applies to the *end time* - the start time needs to be "not greater" (or 0 if that would be negative). Since we'd need two functions or another argument but each is only used once, maybe it's better to do it inline with a comment? Or perhaps to keep _s_to_us that just does the multiplication?

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+        self.period_samples = period_samples
+        self.period_us = self.pelt_sample_us * period_samples
+
+        if duty_cycle_pct:
```
use `if duty_cycle_pct is not None` so that it can be 0.

Commit:
sched: Add hypothesis-based tests for PELT simulator

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+    end_s     : end time [s] used for PELT signal simulation
+    pelt_avg  : expected average value of the PELT signal
+    pelt_init : initial PELT value
+    halt_life : time [ms] of the PELT's halt-life parameter
```
s/halt/half/g

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+
+_PELTRange = namedtuple('PELTRange', [
+    'min_value', 'max_value'])
+class PELTRange(_PELTRange):
```
Love the inheritance trick :tongue: 

I think you could even avoid the `_PELTRange` name by putting the `namedtuple` call inside the brackets. (Not asking you to... just interesting)

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+        - Min stable value: y^i * (1-y^r) / (1-y^p)
+        Where:
+            r is the run time of the task in number of PELT samples
+            p is the period of the task in number of PETL samples
```
s/PETL/PELT/

bjackman commented on bart/sched/pelt.py:
```diff
@@ -0,0 +1,492 @@
+                                           Simulator._sample_us)
+
+            # Assume the task was running for all the current PELT sample
+            active_us = Simulator._sample_us if running else 0
```
Just realised: Should use `self._sample_us` here and elsewhere, then it can be overridden by subclasses or specific instances (also then no explicit references to the class name in the body). No need to set it in `__init__`, attribute lookups fall back to class attributes if there's no instance attribute.

bjackman reviewed (changes requested):
Some nitpicks. 

Also I did some more playing around with Hypothesis: https://github.com/bjackman/bart/blob/pelt_simulator_test/tests/pelt.py 

It does seem like a pretty useful approach to testing, it found a couple of bugs (mentioned in my comments). None of them were very interesting bugs but that test also doesn't test anything very interesting yet.

sinkap commented:
I would like to save this effort from losing the momentum we got at the Plumbers Conference. I think we should merge the changes with the comments extracted in an issue. We do need tests for the code as well @derkling @bjackman can you comment on what is the plan here?

mdigiorgio committed:
sched: add stable range time to PELT simulator

Along with min and max values in the stable range, the stableRange() method now
returns the instant of time when the signal starts to become stable.

mdigiorgio commented:
@derkling could please add these commits b798a37, 86910f2 to the PR? They address comments above from @bjackman and add:

- a feature to the `stableRange()` method to also return the instant of time when the signal starts to become stable
- a method to estimate the initial value of a PELT signal based on when a task starts and the first `sched_load_avg_*` event related to that task

Hope this will create some momentum again in order to get the simulator merged.

mdigiorgio committed:
sched: add PELT initial value estimation method

derkling commented:
@sinkap and @bjackman here are some new patches from @mdigiorgio which should fix all the remaining comments. Can you please have a look and merge, we have some new LISA tests which depends on these bits.

mdigiorgio commented:
@derkling I've updated the two commits, could you please re-push them here? https://github.com/mdigiorgio/bart/commits/pelt_simulator

I'll try to add tests asap.

mdigiorgio commented:
@derkling could you please update the PR with the commits in https://github.com/mdigiorgio/bart/tree/pelt_simulator

Main changes are:
- added tests developed by @bjackman (`hypothesis` module is pretty cool)
- discovered and fixed a bug in `getSignal` thanks to the tests
- few other minor changes

mdigiorgio committed:
sched: add mean value test for PELT simulator

derkling commented:
@sinkap just pushed the new patches from @mdigiorgio .... which I forgot this morning when we pinged you ;-)

sinkap commented:
I am reviewing this code. Sorry about the delay. I will have it done within this week.

sinkap commented:
Merged!

sinkap commented:
Can we update:

https://docs.google.com/document/d/12OmbtpVi_qyBWpFdEMWZ77uuA4A9nN9eGtXuehbeJ-Y/edit

To reflect the reality now that this is merged please?
